# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [2]:
path=os.getcwd()
path

'C:\\Users\\MONIKA\\py3\\Scripts'

In [3]:
path="D:\SPRINGBOARD_1\GuidedCapstone\Step 2\Step 2\data"
os.chdir(path)
os.listdir(path)

['model_performance_metrics.csv', 'step2_output.csv', 'step3_output.csv']

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [4]:
df=pd.read_csv(os.path.join(path,r'step3_output.csv') )
df.head(5)

Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  SkiableTerrain_ac  \
0        0.0          0          2     2  ...             1610.0   
1        0.0          0          0     0  ...              640.0   
2        0.0          0          0     0  ...               30.0   
3        0.0          1          0     2  ...              777.0   
4        0.0          0          1     2  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         1  
1          53.0               90.0             0.0         1  
2          34.0              152.0            30.0         1  
3          89.0              122.0             0.0         0  
4          78.0              104.0            80.0         0  

[5 rows x 27 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [5]:
dfo=df['state']
dfo

0       Alaska
1       Alaska
2       Alaska
3      Arizona
4      Arizona
        ...   
325    Wyoming
326    Wyoming
327    Wyoming
328    Wyoming
329    Wyoming
Name: state, Length: 330, dtype: object

In [6]:
pd.get_dummies(dfo)

Alaska  Arizona  California  Colorado  Connecticut  Idaho  Illinois  \
0         1        0           0         0            0      0         0   
1         1        0           0         0            0      0         0   
2         1        0           0         0            0      0         0   
3         0        1           0         0            0      0         0   
4         0        1           0         0            0      0         0   
..      ...      ...         ...       ...          ...    ...       ...   
325       0        0           0         0            0      0         0   
326       0        0           0         0            0      0         0   
327       0        0           0         0            0      0         0   
328       0        0           0         0            0      0         0   
329       0        0           0         0            0      0         0   

     Indiana  Iowa  Maine  ...  Rhode Island  South Dakota  Tennessee  Utah  \
0          0     0      0  ...             0             0          0     0   
1          0     0      0  ...             0             0          0     0   
2          0     0      0  ...             0             0          0     0   
3          0     0      0  ...             0             0          0     0   
4          0     0      0  ...             0             0          0     0   
..       ...   ...    ...  ...           ...           ...        ...   ...   
325        0     0      0  ...             0             0          0     0   
326        0     0      0  ...             0             0          0     0   
327        0     0      0  ...             0             0          0     0   
328        0     0      0  ...             0             0          0     0   
329        0     0      0  ...             0             0          0     0   

     Vermont  Virginia  Washington  West Virginia  Wisconsin  Wyoming  
0          0         0           0              0          0        0  
1          0         0           0              0          0        0  
2          0         0           0              0          0        0  
3          0         0           0              0          0        0  
4          0         0           0              0          0        0  
..       ...       ...         ...            ...        ...      ...  
325        0         0           0              0          0        1  
326        0         0           0              0          0        1  
327        0         0           0              0          0        1  
328        0         0           0              0          0        1  
329        0         0           0              0          0        1  

[330 rows x 35 columns]

In [7]:
df=df.drop(['state'],axis=1)
df.head(5)

Name  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort         3939           2500        250      1   
1  Eaglecrest Ski Area         2600           1540       1200      0   
2     Hilltop Ski Area         2090            294       1796      0   
3     Arizona Snowbowl        11500           2300       9200      0   
4  Sunrise Park Resort        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  SkiableTerrain_ac  \
0        0.0          0          2     2       0  ...             1610.0   
1        0.0          0          0     0       0  ...              640.0   
2        0.0          0          0     0       1  ...               30.0   
3        0.0          1          0     2       2  ...              777.0   
4        0.0          0          1     2       3  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         1  
1          53.0               90.0             0.0         1  
2          34.0              152.0            30.0         1  
3          89.0              122.0             0.0         0  
4          78.0              104.0            80.0         0  

[5 rows x 26 columns]

In [8]:
df = pd.concat([df, pd.get_dummies(dfo)], axis=1)
df.head(5)

Name  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort         3939           2500        250      1   
1  Eaglecrest Ski Area         2600           1540       1200      0   
2     Hilltop Ski Area         2090            294       1796      0   
3     Arizona Snowbowl        11500           2300       9200      0   
4  Sunrise Park Resort        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  Rhode Island  \
0        0.0          0          2     2       0  ...             0   
1        0.0          0          0     0       0  ...             0   
2        0.0          0          0     0       1  ...             0   
3        0.0          1          0     2       2  ...             0   
4        0.0          0          1     2       3  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 61 columns]

In [9]:
df.T

0                    1                 2    \
Name           Alyeska Resort  Eaglecrest Ski Area  Hilltop Ski Area   
summit_elev              3939                 2600              2090   
vertical_drop            2500                 1540               294   
base_elev                 250                 1200              1796   
trams                       1                    0                 0   
...                       ...                  ...               ...   
Virginia                    0                    0                 0   
Washington                  0                    0                 0   
West Virginia               0                    0                 0   
Wisconsin                   0                    0                 0   
Wyoming                     0                    0                 0   

                            3                    4    \
Name           Arizona Snowbowl  Sunrise Park Resort   
summit_elev               11500                11100   
vertical_drop              2300                 1800   
base_elev                  9200                 9200   
trams                         0                    0   
...                         ...                  ...   
Virginia                      0                    0   
Washington                    0                    0   
West Virginia                 0                    0   
Wisconsin                     0                    0   
Wyoming                       0                    0   

                                         5              6            7    \
Name           Yosemite Ski & Snowboard Area  Bear Mountain  Bear Valley   
summit_elev                             7800           8805         8500   
vertical_drop                            600           1665         1900   
base_elev                               7200           7140         6600   
trams                                      0              0            0   
...                                      ...            ...          ...   
Virginia                                   0              0            0   
Washington                                 0              0            0   
West Virginia                              0              0            0   
Wisconsin                                  0              0            0   
Wyoming                                    0              0            0   

                                  8            9    ...                320  \
Name           Boreal Mountain Resort  Dodge Ridge  ...  Whitecap Mountain   
summit_elev                      7700         8200  ...               1750   
vertical_drop                     500         1600  ...                400   
base_elev                        7200         6600  ...               1295   
trams                               0            0  ...                  0   
...                               ...          ...  ...                ...   
Virginia                            0            0  ...                  0   
Washington                          0            0  ...                  0   
West Virginia                       0            0  ...                  0   
Wisconsin                           0            0  ...                  1   
Wyoming                             0            0  ...                  0   

                           321                   322            323  \
Name           Wilmot Mountain  Grand Targhee Resort  Hogadon Basin   
summit_elev               1030                  9920           8000   
vertical_drop              230                  2270            640   
base_elev                  800                  7851           7400   
trams                        0                     0              0   
...                        ...                   ...            ...   
Virginia                     0                     0              0   
Washington                   0                     0              0   

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [10]:
#---- StandardScaler().fit(data).transform(data)

# It is used for the purpose of feature scaling so that all elements fall in the same range , 
# i.e , It calculates the standardised value for each element Xi of a column as follows :
#          Xi = (Xi- μ) / σ
#      where   μ=Mean of all elements of the column 
#              σ=Std Dev of all elements of the column 
# https://www.youtube.com/watch?v=PiDfEo7qpcw

In [11]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend'] 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

print("X---",X,"y---",y,"scaler---",scaler,"X_scaled---",X_scaled)

X---      summit_elev  vertical_drop  base_elev  trams  fastEight  fastSixes  \
0           3939           2500        250      1        0.0          0   
1           2600           1540       1200      0        0.0          0   
2           2090            294       1796      0        0.0          0   
3          11500           2300       9200      0        0.0          1   
4          11100           1800       9200      0        0.0          0   
..           ...            ...        ...    ...        ...        ...   
325         9500           1000       8500      0        0.0          0   
326         7428            810       6619      0        0.0          0   
327         7808           1571       6237      0        0.0          0   
328         9663            990       8798      0        0.0          0   
329         9500           1100       8400      0        0.0          0   

     fastQuads  quad  triple  double  ...  Rhode Island  South Dakota  \
0            2     2 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [12]:
# --- test_size is the size of Test data

#  Random_state can be 0 or 1 or any other integer. It should be the same value if you want to validate your processing over 
# multiple runs of the code.By the way, I have seen random_state=42 used in many official examples of scikit.The random_state parameter 
# is used for initializing the internal random number generator, which will decide the splitting of data into 
# train and test indices in your case.If random_state is None or np.random, then a randomly-initialized RandomState object is 
# returned.If random_state is an integer, then it is used to seed a new RandomState object.This is to check and validate 
# the data when running the code multiple times. Setting random_state a fixed value will guarantee that the same sequence 
# of random numbers is generated each time you run the code.

In [13]:
# --- Splitting into train , test and validation sets requires using train_test_split twice :
# X_train, X_test, y_train, y_test 
#     = train_test_split(X, y, test_size=0.2, random_state=1)

#  X_train, X_val, y_train, y_val 
#     = train_test_split(X_train, y_train, test_size=0.25, random_state=1) # 0.25 x 0.8 = 0.2

In [14]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 

X_train, X_test,y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [15]:
y

array([ 85.        ,  53.        ,  34.        ,  89.        ,
        78.        ,  47.        ,  64.16681004,  64.16681004,
        64.16681004,  78.        ,  75.        ,  64.16681004,
        64.16681004,  64.16681004, 159.        ,  59.        ,
        84.        ,  69.        ,  64.16681004,  64.16681004,
        83.        ,  64.16681004,  79.        ,  50.        ,
       125.        ,  69.        ,  85.        , 179.        ,
        64.16681004,  64.16681004, 158.        ,  64.16681004,
        89.        ,  64.16681004,  25.        ,  64.16681004,
        79.        ,  89.        ,  71.        ,  79.        ,
        56.        ,  84.        ,  64.16681004,  65.        ,
       139.        ,  64.16681004,  64.16681004,  72.        ,
        65.        ,  60.        ,  55.        ,  62.        ,
        42.        ,  64.        ,  70.        ,  42.        ,
        47.        ,  32.        ,  47.        ,  64.16681004,
        81.        ,  62.        ,  43.        ,  64.16

In [16]:
print("X_train---",X_train,"\nX_test---",X_test,"\ny_train---",y_train,"\ny_test---" ,y_test)

X_train--- [[-0.15866985 -0.91441494  0.04048329 ... -0.11076976 -0.22573306
  -0.15762208]
 [-0.25518791  1.47351277 -0.75311775 ... -0.11076976 -0.22573306
  -0.15762208]
 [-0.97639234 -0.9672452  -0.87521021 ... -0.11076976 -0.22573306
  -0.15762208]
 ...
 [-0.96968969 -0.99366033 -0.82701582 ... -0.11076976 -0.22573306
  -0.15762208]
 [ 1.05853018  2.61464637  0.84372321 ... -0.11076976 -0.22573306
  -0.15762208]
 [ 1.97330695 -0.05645153  2.38273089 ... -0.11076976 -0.22573306
  -0.15762208]] 
X_test--- [[ 1.18185882 -0.22762157  1.48631514 ... -0.11076976 -0.22573306
  -0.15762208]
 [ 1.29499944  1.38053152  1.13288958 ... -0.11076976 -0.22573306
  -0.15762208]
 [ 1.85212316  0.72332309  2.00038869 ... -0.11076976 -0.22573306
  -0.15762208]
 ...
 [-0.99757269 -1.03064151 -0.88131484 ... -0.11076976 -0.22573306
  -0.15762208]
 [-0.70292449 -0.59743339 -0.65994192 ... -0.11076976 -0.22573306
  -0.15762208]
 [ 0.6992685   0.1950205   0.77946401 ... -0.11076976 -0.22573306
  -0.15762

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [17]:
# all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [18]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred=model.predict(X_test)
print("X_test",X_test,"\ny_pred",y_pred )


X_test [[ 1.18185882 -0.22762157  1.48631514 ... -0.11076976 -0.22573306
  -0.15762208]
 [ 1.29499944  1.38053152  1.13288958 ... -0.11076976 -0.22573306
  -0.15762208]
 [ 1.85212316  0.72332309  2.00038869 ... -0.11076976 -0.22573306
  -0.15762208]
 ...
 [-0.99757269 -1.03064151 -0.88131484 ... -0.11076976 -0.22573306
  -0.15762208]
 [-0.70292449 -0.59743339 -0.65994192 ... -0.11076976 -0.22573306
  -0.15762208]
 [ 0.6992685   0.1950205   0.77946401 ... -0.11076976 -0.22573306
  -0.15762208]] 
y_pred [ 57.97414808  91.28762464  78.7236598   76.54055433  69.90090589
  76.85598402  45.11623792  46.94094495  93.67531995  83.45656995
  51.10842542  79.0464137   51.0698512   71.58645277  62.8276637
  92.37942152  36.91653089  72.68508558  61.5386012   41.35989027
 113.64895277  36.67531995  53.18117933  49.15188245  48.6089137
  59.99367933  77.43996839  75.66946058  64.09035902  66.24148206
  35.52492933  61.0932887   79.8589137   56.81545667  75.35403089
  49.25442152  23.16848402  28.91

## Review Model Outcomes — Iterate over additional models as needed

In [19]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
from sklearn.metrics import explained_variance_score , mean_absolute_error
# You can plug y_test and y_pred into the functions to evaluate the model
var_score=explained_variance_score(y_test , y_pred )
mean_abs_err=mean_absolute_error(y_test , y_pred)


In [20]:
print("var_score---",var_score,"mean_abs_err---",mean_abs_err)

var_score--- 0.9314755063573666 mean_abs_err--- 5.141424676548993


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [21]:
print(model.intercept_)

64.1186793286615


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [22]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coeff=pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
coeff

Coefficient
summit_elev        2.086137e+01
vertical_drop      6.539289e+00
base_elev          1.905014e+01
trams              6.481119e+11
fastEight          6.371575e+10
fastSixes          7.542958e+11
fastQuads          2.544424e+12
quad               1.518863e+12
triple             1.874068e+12
double             2.100811e+12
surface            2.383934e+12
total_chairs       6.711709e+12
Runs               1.283418e+00
TerrainParks       2.014075e-01
LongestRun_mi      3.077259e-01
SkiableTerrain_ac  2.198624e-01
Snow Making_ac     3.123537e-01
daysOpenLastYear   2.643066e-01
yearsOpen          3.884201e-01
averageSnowfall    3.361422e-01
AdultWeekday       1.919806e+01
projectedDaysOpen  9.357150e-02
NightSkiing_ac     8.434156e-01
clusters           1.087073e+00
Alaska             1.093073e+12
Arizona            8.938541e+11
California         2.811277e+12
Colorado           2.872774e+12
Connecticut        1.406829e+12
Idaho              2.155852e+12
Illinois           1.260241e+12
Indiana            8.938541e+11
Iowa               1.093073e+12
Maine              1.875808e+12
Maryland           6.330131e+11
Massachusetts      2.067314e+12
Michigan           3.260596e+12
Minnesota          2.321252e+12
Missouri           8.938541e+11
Montana            2.155852e+12
Nevada             1.260241e+12
New Hampshire      2.473657e+12
New Jersey         8.938541e+11
New Mexico         1.875808e+12
New York           3.455015e+12
North Carolina     1.538732e+12
Ohio               1.406829e+12
Oregon             1.974193e+12
Pennsylvania       2.682698e+12
Rhode Island       6.330131e+11
South Dakota       8.938541e+11
Tennessee          6.330131e+11
Utah               2.240351e+12
Vermont            2.398919e+12
Virginia           1.260241e+12
Washington         1.974193e+12
West Virginia      1.260241e+12
Wisconsin          2.473657e+12
Wyoming            1.771282e+12

In [23]:
coeff.sort_values(['Coefficient']).tail(10)

Coefficient
Vermont        2.398919e+12
Wisconsin      2.473657e+12
New Hampshire  2.473657e+12
fastQuads      2.544424e+12
Pennsylvania   2.682698e+12
California     2.811277e+12
Colorado       2.872774e+12
Michigan       3.260596e+12
New York       3.455015e+12
total_chairs   6.711709e+12

In [24]:
toptenCoeff=coeff.sort_values(['Coefficient'],ascending=False)
toptenCoeff.head(10)

Coefficient
total_chairs   6.711709e+12
New York       3.455015e+12
Michigan       3.260596e+12
Colorado       2.872774e+12
California     2.811277e+12
Pennsylvania   2.682698e+12
fastQuads      2.544424e+12
New Hampshire  2.473657e+12
Wisconsin      2.473657e+12
Vermont        2.398919e+12

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [25]:
df2=pd.read_csv(os.path.join(path,r'step3_output.csv') )
df2.head(5)

Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  SkiableTerrain_ac  \
0        0.0          0          2     2  ...             1610.0   
1        0.0          0          0     0  ...              640.0   
2        0.0          0          0     0  ...               30.0   
3        0.0          1          0     2  ...              777.0   
4        0.0          0          1     2  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         1  
1          53.0               90.0             0.0         1  
2          34.0              152.0            30.0         1  
3          89.0              122.0             0.0         0  
4          78.0              104.0            80.0         0  

[5 rows x 27 columns]

In [26]:
df2.drop(['state'],axis=1,inplace = True)
df2.head()

Name  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort         3939           2500        250      1   
1  Eaglecrest Ski Area         2600           1540       1200      0   
2     Hilltop Ski Area         2090            294       1796      0   
3     Arizona Snowbowl        11500           2300       9200      0   
4  Sunrise Park Resort        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  SkiableTerrain_ac  \
0        0.0          0          2     2       0  ...             1610.0   
1        0.0          0          0     0       0  ...              640.0   
2        0.0          0          0     0       1  ...               30.0   
3        0.0          1          0     2       2  ...              777.0   
4        0.0          0          1     2       3  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         1  
1          53.0               90.0             0.0         1  
2          34.0              152.0            30.0         1  
3          89.0              122.0             0.0         0  
4          78.0              104.0            80.0         0  

[5 rows x 26 columns]

In [27]:
# STANDARDIZING

X2 = df2.drop(['Name','AdultWeekend'], axis=1) 
y2 = df2['AdultWeekend'] 
scaler2 = preprocessing.StandardScaler().fit(X2)
X_scaled2=scaler2.transform(X2) 

print("X---",X2,"y---",y2,"scaler---",scaler2,"X_scaled---",X_scaled2)

X---      summit_elev  vertical_drop  base_elev  trams  fastEight  fastSixes  \
0           3939           2500        250      1        0.0          0   
1           2600           1540       1200      0        0.0          0   
2           2090            294       1796      0        0.0          0   
3          11500           2300       9200      0        0.0          1   
4          11100           1800       9200      0        0.0          0   
..           ...            ...        ...    ...        ...        ...   
325         9500           1000       8500      0        0.0          0   
326         7428            810       6619      0        0.0          0   
327         7808           1571       6237      0        0.0          0   
328         9663            990       8798      0        0.0          0   
329         9500           1100       8400      0        0.0          0   

     fastQuads  quad  triple  double  ...  LongestRun_mi  SkiableTerrain_ac  \
0            2 

In [28]:
# TEST TRAIN SPLIT 

y2 = y2.ravel()
X_train2, X_test2,y_train2, y_test2 = train_test_split(X_scaled2, y2, test_size=0.25, random_state=1)

In [29]:
#LINEAR REGRESSION 
model2 = lm.fit(X_train2,y_train2)
y_pred2=model2.predict(X_test2)
print("X_test",X_test2,"\ny_pred",y_pred2 )

X_test [[ 1.18185882e+00 -2.27621573e-01  1.48631514e+00 ...  6.25344110e-01
   4.61933464e-01  1.53245504e+00]
 [ 1.29499944e+00  1.38053152e+00  1.13288958e+00 ...  9.04095649e-01
   2.60502419e+00 -1.68862880e+00]
 [ 1.85212316e+00  7.23323092e-01  2.00038869e+00 ...  1.37528917e-01
  -6.09611897e-01 -1.68862880e+00]
 ...
 [-9.97572690e-01 -1.03064151e+00 -8.81314836e-01 ... -3.50286276e-01
   3.01201660e-01 -7.80868809e-02]
 [-7.02924488e-01 -5.97433387e-01 -6.59941915e-01 ... -6.98725699e-01
   2.47624392e-01 -7.80868809e-02]
 [ 6.99268501e-01  1.95020500e-01  7.79464015e-01 ... -9.90324405e-16
  -6.09611897e-01  1.53245504e+00]] 
y_pred [ 57.07744563  86.07167181  82.90550928  78.427293    66.83457283
  82.058798    38.05033188  47.8261433   88.9377061   85.51580197
  52.04041694  80.90394404  52.93653946  65.7895346   63.43525827
  88.32831875  37.55594469  77.86794846  63.30755142  43.46694728
 109.24413667  37.34736695  53.88640976  50.14428103  50.59616187
  58.67340017  66.6

In [30]:
# VARIANCE SCORE AND ERROR
var_score2=explained_variance_score(y_test2 , y_pred2 )
mean_abs_err2=mean_absolute_error(y_test2 , y_pred2)
print("var_score2---",var_score2,"\nmean_abs_err2---",mean_abs_err2)

var_score2--- 0.9240544500362643 
mean_abs_err2--- 5.491640284244614


In [31]:
# INTERCEPT
print(model2.intercept_)

64.06214953957362


In [32]:
#COEFFICIENTS
coeff2=pd.DataFrame(abs(lm.coef_), X2.columns, columns=['Coefficient'])
coeff2=coeff2.sort_values(['Coefficient'],ascending=False)
coeff2.head(10)

Coefficient
AdultWeekday        19.999313
summit_elev         13.619945
base_elev           10.442217
vertical_drop        4.873477
averageSnowfall      1.582491
quad                 1.561263
Runs                 1.486680
triple               1.447902
surface              1.311456
daysOpenLastYear     1.021380

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [33]:
df3=pd.read_csv(os.path.join(path,r'step3_output.csv') )
df3.head(5)

Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  SkiableTerrain_ac  \
0        0.0          0          2     2  ...             1610.0   
1        0.0          0          0     0  ...              640.0   
2        0.0          0          0     0  ...               30.0   
3        0.0          1          0     2  ...              777.0   
4        0.0          0          1     2  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         1  
1          53.0               90.0             0.0         1  
2          34.0              152.0            30.0         1  
3          89.0              122.0             0.0         0  
4          78.0              104.0            80.0         0  

[5 rows x 27 columns]

In [34]:
df3.drop(['state','summit_elev'],axis=1,inplace=True)
df3.columns

Index(['Name', 'vertical_drop', 'base_elev', 'trams', 'fastEight', 'fastSixes',
       'fastQuads', 'quad', 'triple', 'double', 'surface', 'total_chairs',
       'Runs', 'TerrainParks', 'LongestRun_mi', 'SkiableTerrain_ac',
       'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen', 'averageSnowfall',
       'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen', 'NightSkiing_ac',
       'clusters'],
      dtype='object')

In [35]:
#STANDARDIZING
X3 = df3.drop(['Name','AdultWeekend'], axis=1) 
y3 = df3['AdultWeekend'] 
scaler3 = preprocessing.StandardScaler().fit(X3)
X_scaled3=scaler3.transform(X3) 

# print("X---",X3,"y---",y3,"scaler---",scaler3,"X_scaled---",X_scaled3)

In [36]:
# TEST TRAIN SPLIT 

y3 = y3.ravel()
X_train3, X_test3,y_train3, y_test3 = train_test_split(X_scaled3, y3, test_size=0.25, random_state=1)

In [37]:
#LINEAR REGRESSION 
model3 = lm.fit(X_train3,y_train3)
y_pred3=model3.predict(X_test3)
# print("X_test",X_test3,"\ny_pred",y_pred3 )

In [38]:
# VARIANCE SCORE AND ERROR
var_score3=explained_variance_score(y_test3 , y_pred3 )
mean_abs_err3=mean_absolute_error(y_test3 , y_pred3)
print("var_score3---",var_score3,"\nmean_abs_err3---",mean_abs_err3)

var_score3--- 0.9300129303202191 
mean_abs_err3--- 5.291954940892436


In [39]:
#INTERCEPT
print(model3.intercept_)

64.06467965536973


In [40]:
#COEFFICIENTS
coeff3=pd.DataFrame(abs(lm.coef_), X3.columns, columns=['Coefficient'])
coeff3=coeff3.sort_values(['Coefficient'],ascending=False)
coeff3.head(10)

Coefficient
AdultWeekday        19.971353
vertical_drop        1.600656
Runs                 1.593790
averageSnowfall      1.543505
quad                 1.516445
triple               1.427518
surface              1.253857
daysOpenLastYear     1.069929
base_elev            1.009642
clusters             0.907752

In [41]:
print("var_score---",var_score,"mean_abs_err---",mean_abs_err)
print("\nvar_score2---",var_score2,"mean_abs_err2---",mean_abs_err2)
print("\nvar_score3---",var_score3,"mean_abs_err3---",mean_abs_err3)

var_score--- 0.9314755063573666 mean_abs_err--- 5.141424676548993

var_score2--- 0.9240544500362643 mean_abs_err2--- 5.491640284244614

var_score3--- 0.9300129303202191 mean_abs_err3--- 5.291954940892436


## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.9314755063573666|5.141424676548993  |-|
| Model 2. | 0.9240544500362643|5.491640284244614 |'state'|
| Model 3. | 0.9300129303202191 |5.291954940892436 |'state','summit_elev','base_elev'|

Model Selection:

Model 1 is the best model since its 
-MAE is minimum in comparison to Model 2 and Model 3 and
-Variance score is near 1.0 i.e. greater compared to other models